From 3427910cd2a840a8b3a8c3a1f372fd4656e49e34 Mon Sep 17 00:00:00 2001
From: Shrinivas Prabhu <shrinivasx.prabhu@intel.com>
Date: Wed, 11 Jul 2018 23:39:09 +0530
Subject: [PATCH] Handling multiple ethernet interfaces

This fix handles the multiple ethernet interfaces when the current
interface is in link down state.

Change-Id: Iff6c84cc8c68e42bcfa3f4effbfc42b40dc736b2
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-63367
Signed-off-by: Shrinivas Prabhu <shrinivasx.prabhu@intel.com>
---
 .../android/server/ethernet/EthernetNetworkFactory.java   | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/java/com/android/server/ethernet/EthernetNetworkFactory.java b/java/com/android/server/ethernet/EthernetNetworkFactory.java
index d6d0def..05e8eeb 100644
--- a/java/com/android/server/ethernet/EthernetNetworkFactory.java
+++ b/java/com/android/server/ethernet/EthernetNetworkFactory.java
@@ -50,7 +50,9 @@ import com.android.server.net.BaseNetworkObserver;
 import java.io.FileDescriptor;
 import java.io.PrintWriter;
 import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.concurrent.CountDownLatch;
+import java.util.Collections;
 
 
 /**
@@ -186,6 +188,11 @@ class EthernetNetworkFactory {
         @Override
         public void interfaceAdded(String iface) {
             mHandler.post(() -> {
+                if (isTrackingInterface()) {
+                    if (!mLinkUp) {
+                        stopTrackingInterface(mIface);
+                    }
+                }
                 maybeTrackInterface(iface);
             });
         }
@@ -422,7 +429,15 @@ class EthernetNetworkFactory {
     public void trackFirstAvailableInterface() {
         try {
             final String[] ifaces = mNMService.listInterfaces();
+            ArrayList<String> eth_ifaces = new ArrayList<String>();
             for (String iface : ifaces) {
+                if (iface.matches(mIfaceMatch)) {
+                     eth_ifaces.add(iface);
+                }
+            }
+            Collections.sort(eth_ifaces, Collections.reverseOrder());
+            for (String iface : eth_ifaces) {
+
                 if (maybeTrackInterface(iface)) {
                     // We have our interface. Track it.
                     // Note: if the interface already has link (e.g., if we crashed and got
-- 
1.9.1

